src/coreclr/jit/helperexpansion.cpp

```diff
@@ -478,36 +478,75 @@ bool Compiler::fgExpandThreadLocalAccessForCall(BasicBlock** pBlock, Statement*
         return false;
     }
 
+    assert(!opts.IsReadyToRun());
+
+    if (TargetOS::IsUnix)
+    {
+#if defined(TARGET_ARM) || !defined(TARGET_64BIT)
+        // On Arm, Thread execution blocks are accessed using co-processor registers and instructions such
+        // as MRC and MCR are used to access them. We do not support them and so should never optimize the
+        // field access using TLS.
+        assert(!"Unsupported scenario of optimizing TLS access on Linux Arm32/x86");
+#endif
+    }
+    else
+    {
 #ifdef TARGET_ARM
-    // On Arm, Thread execution blocks are accessed using co-processor registers and instructions such
-    // as MRC and MCR are used to access them. We do not support them and so should never optimize the
-    // field access using TLS.
-    assert(!"Unsupported scenario of optimizing TLS access on Arm32");
+        // On Arm, Thread execution blocks are accessed using co-processor registers and instructions such
+        // as MRC and MCR are used to access them. We do not support them and so should never optimize the
+        // field access using TLS.
+        assert(!"Unsupported scenario of optimizing TLS access on Windows Arm32");
 #endif
+    }
 
     JITDUMP("Expanding thread static local access for [%06d] in " FMT_BB ":\n", dspTreeID(call), block->bbNum);
     DISPTREE(call);
     JITDUMP("\n");
+
     bool isGCThreadStatic =
         eeGetHelperNum(call->gtCallMethHnd) == CORINFO_HELP_GETSHARED_GCTHREADSTATIC_BASE_NOCTOR_OPTIMIZED;
 
     CORINFO_THREAD_STATIC_BLOCKS_INFO threadStaticBlocksInfo;
+    memset(&threadStaticBlocksInfo, 0, sizeof(CORINFO_THREAD_STATIC_BLOCKS_INFO));
+
     info.compCompHnd->getThreadLocalStaticBlocksInfo(&threadStaticBlocksInfo, isGCThreadStatic);
 
-    uint32_t offsetOfMaxThreadStaticBlocksVal = 0;
-    uint32_t offsetOfThreadStaticBlocksVal    = 0;
+    if (TargetOS::IsMacOS)
+    {
+        if (threadStaticBlocksInfo.threadVarsSection == 0)
+        {
+            // We possibly compiled coreclr as single file and not .so file.
+            // Do not perform this optimization for it.
```

Comment:
Just curious - have you seen this path to be actually taken on macOS?

Reply:
Not really but I wanted to make sure that we guard it, thinking of which, I think this check is only needed for osx/x64 because that's the one which does the disassembling. osx/arm64 just gets the value from function call.

Reply:
If it works fine for single file on OSX x64, the comment should not be implying that it does not work.

Reply:
I mean I am not sure if we build single file on OSX/x64, the way we have CI leg for linux/x64. But since we have [this check](https://github.com/dotnet/runtime/pull/87082/files#diff-98b4e379205d747e922689cb993f985be190fedb509fd486e99280c842703472R1662-R1666), thought it would be worthwhile adding.

Reply:
I am ok with having these checks, but the comments attached to them should be accurate.

We do build and ship single for OSX/x64.

Reply:
how about this?

```
            // We possibly compiled coreclr as single file and not .dylib file in which case __thread_vars
            // we found was not accurate. Do not perform this optimization for it.
```